core/templates/dev/head/services/SpeechSynthesisChunkerService.js

```diff
@@ -0,0 +1,190 @@
+     * @param {Number} offset - An integer indicating what offset in the text
+     *    to begin the next chunked utterance.
+     * @param {requestCallback} - Callback that is activated when the last
+     *    chunked utterance finishes playing.
+     */
+    var _speechUtteranceChunker = function(utterance, offset, callback) {
+      var newUtterance;
+      var text = (offset !== undefined ?
+          utterance.text.substring(offset) : utterance.text);
+
+      // This regex pattern finds the next string at most 160
+      // characters in length that ends on a punctuation mark in
+      // PUNCTUATION_MARKS_TO_END_CHUNKS.
+      var delimitChunkRegex = new RegExp('^[\\s\\S]{' +
+        Math.floor(CHUNK_LENGTH / 4) + ',' + CHUNK_LENGTH + '}[' +
+        PUNCTUATION_MARKS_TO_END_CHUNKS + ']{1}|^[\\s\\S]{1,' +
+        CHUNK_LENGTH + '}$|^[\\s\\S]{1,' + CHUNK_LENGTH + '} ');
+      var chunkArray = text.match(delimitChunkRegex);
+      if (chunkArray == null ||
+          chunkArray[0] === undefined ||
+          chunkArray[0].length <= 2) {
+        // Call once all text has been spoken.
+        if (callback !== undefined) {
+          callback();
+        }
+        return;
+      }
+      var chunk = chunkArray[0];
+      newUtterance = new SpeechSynthesisUtterance(chunk);
+
+      // Copy properties from the current utterance to the next utterance,
+      // excluding the text being spoken.
+      for (var property in utterance) {
+        if (property !== 'text') {
+          newUtterance[property] = utterance[property];
+        }
+      }
+      newUtterance.onend = function() {
+        if (_speechUtteranceChunker.cancel) {
+          _speechUtteranceChunker.cancel = false;
+          return;
+        }
+        offset += chunk.length;
+        _speechUtteranceChunker(utterance, offset, callback);
+      }
+
+      // IMPORTANT!! Do not remove: Logging the object out fixes some onend
+      // firing issues. Placing the speak invocation inside a callback
+      // fixes ordering and onend issues.
+      console.log(newUtterance);
+      $timeout(function () {
+        speechSynthesis.speak(newUtterance);
+      }, 0);
+    };
+
+    var _convertToSpeakableText = function(html) {
+      elt = $('<div>' + html + '</div>');
+
+      // Convert links into speakable text by extracting the readable value.
+      elt.find('oppia-noninteractive-' + RTE_COMPONENT_NAMES.Link)
+        .replaceWith(function() {
+          if (this.attributes['text-with-value'] !== undefined) {
+            return this.attributes['text-with-value'].textContent
+              .replace(/&quot;/g, '');
+          }
+        }
+      );
+
+      // Convert LaTeX to speakable text.
+      elt.find('oppia-noninteractive-' + RTE_COMPONENT_NAMES.Math)
+        .replaceWith(function() {
+          if (this.attributes['raw_latex-with-value'] !== undefined) {
+            return this.attributes['raw_latex-with-value'].textContent
```

Comment:
Consider pulling this part out into a function and adding karma tests that feed in a raw-latex-with-value string and that check the result after all the replacements? I can see this breaking as new improvements are added.

Reply:
done